### ISSUE TITLE
[Bug][HiCache] NCCL deadlock in `writing_check` due to conditional `all_reduce` under TP>1 with

### ISSUE THREAD

**Issue by jeremyzhang866** [sgl-project/sglang#26921]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Symptom

Running GLM-5.1 FP8 with TP=8 and `--enable-hierarchical-cache --hicache-write-policy write_through`, the service hangs after several hours. Watchdog timeout fires, SIGQUIT kills the process. py-spy shows the scheduler main thread stuck in `writing_check → _all_reduce_attn_groups → all_reduce`.

This is likely the same root cause as #26357, where the observed symptom (high `evictable`, low `available`, scheduler hang) is consistent with the scheduler deadlocking in `writing_check`'s `all_reduce` before reaching eviction code.

### Root Cause

`writing_check` conditionally calls `all_reduce` based on whether the local `ongoing_write_through` dict is empty:

```python
# hiradix_cache.py L774-776
# NOTE: all ranks has the same ongoing_write_through, can skip sync if empty
if len(self.ongoing_write_through) == 0:
    return  # skips all_reduce
```

The comment assumes all ranks have identical `ongoing_write_through`, but this assumption is violated. `loading_check` (L800-813) processes CPU→GPU DMA completion events **independently on each rank without cross-rank synchronization**:

```python
# hiradix_cache.py L807
# no need to sync across TP workers as batch forwarding is synced
for ack_id in ack_list:
    end_node = self.ongoing_load_back.pop(ack_id)
    self.dec_lock_ref(end_node)  # each rank unlocks different nodes at different times
```

DMA completion timing differs across ranks (hardware-level PCIe variation). This causes the following divergence chain:

```
loading_check unsynchronized → lock_ref divergence → evictable_size divergence
→ different evict decisions → different write_backup triggers
→ ongoing_write_through differs across ranks
→ some ranks return early (empty), others call all_reduce (not empty)
→ NCCL op sequence permanently desynchronized → deadlock
```

The divergence accumulates over thousands of iterations. Under high memory pressure (frequent evictions + load_backs), it eventually reaches a critical point where `ongoing_write_through` is empty on some ranks but not others.

### Environment

| Component | Version |
|-----------|---------|
| SGLang | 0.5.4.post1 |
| PyTorch | 2.6.0+cu126 |
| CUDA | 12.6 |
| NCCL | 2.21.5 |
| Python | 3.10.16 |
| GPU | NVIDIA H20 × 8 |
| Driver | 570.133.20 |
| OS | Linux 5.10.134 (Alibaba Cloud) |

**Comment by jeremyzhang866** [comment]
# HiCache NCCL Deadlock — Reproduction Guide

Reproduces the full causal chain in ~27 seconds (vs hours in production):

```
loading_check divergence → lock_ref → evict → write_backup
→ ongoing_write_through divergence → conditional all_reduce mismatch → deadlock
```

> **Note**: Tested on NVIDIA H20 × 2 with Qwen3-0.6B. If deadlock does not trigger on your setup, reduce `mem-fraction-static` (try `0.02`) or `max-total-tokens` (try `1024`) to increase eviction frequency. The injection is deterministic — if `load_back` events occur, divergence will happen. The key is making sure memory pressure is high enough to trigger frequent evictions.

## 1. Inject into loading_check

File: `sglang/srt/mem_cache/hiradix_cache.py`, function `loading_check`

Add two blocks (marked with `+`) to make Rank 1 process at most 1 event per round, then skip 10 rounds:

```python
def loading_check(self):
+   if self.tp_world_size > 1:
+       import torch.distributed as dist
+       if dist.get_rank() != 0:
+           if not hasattr(self, '_load_skip_remaining'):
+               self._load_skip_remaining = 0
+           if self._load_skip_remaining > 0:
+               self._load_skip_remaining -= 1
+               return

    finish_count = 0
    for _, finish_event, ack_list in self.cache_controller.ack_load_queue:
        if not finish_event.query():
            break
        finish_count += 1
        for ack_id in ack_list:
            end_node = self.ongoing_load_back.pop(ack_id)
            self.dec_lock_ref(end_node)

+       if self.tp_world_size > 1:
+           import torch.distributed as dist
+           if dist.get_rank() != 0 and finish_count >= 1:
+               self._load_skip_remaining = 10
+               break

    del self.cache_controller.ack_load_queue[:finish_count]
```

## 2. Start server with minimal KV cache

```bash
python -m sglang.launch_server \
    --model Qwen3-0.6B --tp 2 --port 18899 \
    --enable-hierarchical-cache --hicache-write-policy write_through \
    --hicache-size 50 --mem-fraction-static 0.05 \
    --max-total-tokens 2048 --max-running-requests 16 \
    --disable-cuda-graph --attention-backend triton \
    --decode-attention-backend triton --sampling-backend pytorch \
    --trust-remote-code
```

Key: `mem-fraction-static 0.05` + `max-total-tokens 2048` forces frequent evictions.

## 3. Run stress test

```bash
#!/bin/bash
SERVER="http://localhost:18899"
MODEL="Qwen3-0.6B"

PREFIXES=(
    "The history of artificial intelligence began in the 1950s when pioneers"
    "In distributed computing systems consensus algorithms play a critical role"
    "Machine learning models require careful optimization of hyperparameters"
    "The architecture of modern neural networks has evolved significantly"
    "Database management systems handle concurrent transactions through various"
    "Cloud computing infrastructure relies on virtualization technologies that"
    "Natural language processing has been revolutionized by large language models"
    "The principles of software engineering emphasize modularity abstraction"
)

send() {
    timeout 30 curl -s "$SERVER/v1/chat/completions" \
        -H "Content-Type: application/json" \
        -d "{\"model\":\"$MODEL\",\"messages\":[{\"role\":\"user\",\"content\":\"$1\"}],\"max_tokens\":64}" \
        > /dev/null 2>&1 &
}

# Wait for server
while ! curl -s -o /dev/null -w "%{http_code}" "$SERVER/health" | grep -q 200; do sleep 2; done

round=0
while true; do
    round=$((round + 1))
    echo "=== Round $round ==="

    # Phase 1: shared prefix → fill cache + write_through backup to CPU
    for i in $(seq 0 7); do send "${PREFIXES[$i]} explain $round $RANDOM"; done; wait

    # Phase 2: diverse requests → GPU fills up → evictions
    for i in $(seq 1 16); do send "Random query $round $i $RANDOM $RANDOM explain"; done; wait

    # Phase 3: re-send shared prefix → cache miss GPU, hit CPU → load_back
    for i in $(seq 0 7); do send "${PREFIXES[$i]} different angle $round $RANDOM"; done; wait

    # Check health
    curl -s -o /dev/null -w "%{http_code}" "$SERVER/health" | grep -q 200 || { echo "DEADLOCK at round $round"; break; }
    sleep 0.5
done
```

Phase 3 is critical — it triggers `load_back` (CPU→GPU), which gives `loading_check` events to process, making the injection effective.

## 4. Result

Deadlock in ~27 seconds. Server returns 503, watchdog timeout after 300s.

```
evictable_size: Rank 0=402, Rank 1=165 (diverged)
py-spy: TP0 stuck at broadcast, TP1 stuck at all_reduce (NCCL desync)
```

## 5. Verify fix

Apply fix in `writing_check` (make `all_reduce` unconditional), keep injection + same params:

```diff
-        if len(self.ongoing_write_through) == 0:
-            return
         finish_count = 0
-        for _, finish_event, ack_list in ...:
+        if len(self.ongoing_write_through) > 0:
+            for _, finish_event, ack_list in ...:
```

Result: 50 rounds, 1419 requests, zero deadlock.

**Comment by NoneGG** [comment]
Any update here?

**Comment by jeremyzhang866** [comment]
> Any update here?

The proposed fix is in #26923.  
It has been validated on my side, and we are waiting for maintainer/committer confirmation.
